rollup/internal/controller/relayer/l2_relayer.go

```diff
@@ -373,11 +373,178 @@ func (r *Layer2Relayer) ProcessGasPriceOracle() {
 // ProcessPendingBatches processes the pending batches by sending commitBatch transactions to layer 1.
 func (r *Layer2Relayer) ProcessPendingBatches() {
 	// get pending batches from database in ascending order by their index.
-	dbBatches, err := r.batchOrm.GetFailedAndPendingBatches(r.ctx, 5)
+	dbBatches, err := r.batchOrm.GetFailedAndPendingBatches(r.ctx, max(5, r.cfg.SenderConfig.BatchSubmission.MaxBatches))
 	if err != nil {
 		log.Error("Failed to fetch pending L2 batches", "err", err)
 		return
 	}
+
+	var batchesToSubmit []*dbBatchWithChunksAndParent
+	var forceSubmit bool
+	for i, dbBatch := range dbBatches {
+		if i == 0 && encoding.CodecVersion(dbBatch.CodecVersion) < encoding.CodecV6 {
+			// if the first batch is not >= V6 then we need to submit batches one by one
+			r.processPendingBatchesV4(dbBatches)
+			return
+		}
+
+		batchesToSubmitLen := len(batchesToSubmit)
+		var dbChunks []*orm.Chunk
+		var dbParentBatch *orm.Batch
+
+		// Verify batches compatibility
+		{
+			dbChunks, err = r.chunkOrm.GetChunksInRange(r.ctx, dbBatch.StartChunkIndex, dbBatch.EndChunkIndex)
+			if err != nil {
+				log.Error("failed to get chunks in range", "err", err)
+				return
+			}
+
+			// check codec version
+			for _, dbChunk := range dbChunks {
+				if dbBatch.CodecVersion != dbChunk.CodecVersion {
+					log.Error("batch codec version is different from chunk codec version", "batch index", dbBatch.Index, "chunk index", dbChunk.Index, "batch codec version", dbBatch.CodecVersion, "chunk codec version", dbChunk.CodecVersion)
+					return
+				}
+			}
+
+			if dbBatch.Index == 0 {
+				log.Error("invalid args: batch index is 0, should only happen in committing genesis batch")
+				return
+			}
+
+			dbParentBatch, err = r.batchOrm.GetBatchByIndex(r.ctx, dbBatch.Index-1)
```

Comment:
Most of these batches are already in memory (`dbBatches`).

Reply:
changed to use `dbBatches` except for `i==0` in 5a479c3a089f4a27b52be5461549cc19504625fe